### PR TITLE
Remove `boost` dependency

### PR DESCRIPTION
**Pull request by codereport**
Once we remove Boost `filesystem` dependency and use `std::filesystem`, we can remove the boost dependency.

This change requires dropping support for GCC < 9.

### CODE REVIEW

**Review by ttnghia — APPROVED** [review]

**Comment by kkraus14** [comment]
Can you remove boost from the conda recipes / environments as well?

- https://github.com/rapidsai/cudf/blob/branch-0.20/conda/recipes/libcudf/meta.yaml#L42
- https://github.com/rapidsai/cudf/blob/branch-0.20/conda/environments/cudf_dev_cuda11.0.yml#L51 (can drop the 10.x environment files while you're at it)
- https://github.com/rapidsai/integration/blob/branch-0.19/conda/recipes/versions.yaml#L35-L36
- https://github.com/rapidsai/integration/blob/branch-0.19/conda/recipes/rapids-build-env/meta.yaml#L46-L47

**Comment by davidwendt** [comment]
Boost is also installed in the docker_build Docker file
https://github.com/rapidsai/cudf/blob/7ad07b49e73ddbaca32e16cf491953a5e98b79bf/docker_build/Dockerfile#L43-L45
I'm not sure if the `libboost-system-dev` and `libboost-regex-dev` are needed for some other purpose.

File: java/LICENSE-bundled
Comment (codereport) [inline java/LICENSE-bundled:4]: @kkraus14 is it ok to remove this LICENSE?

Comment (kkraus14) [reply java/LICENSE-bundled:4]: I would defer to @rapidsai/cudf-java-codeowners

**Comment by codereport** [comment]
rerun tests

File: java/LICENSE-bundled
Comment (harrism) [reply java/LICENSE-bundled:4]: @sameerz @JasonAtNvidia ?

Comment (jlowe) [reply java/LICENSE-bundled:4]: Yes, this LICENSE should be removed if we are no longer linking against boost.  The Java bindings currently are building libcudf requesting a statically linked boost filesystem library to avoid having problems locating the correct version of the library at runtime.

File: java/README.md
Comment (jlowe) [inline java/README.md:71]: Statically linking Arrow is still needed for the Java build.  The cudf jar is not used in a conda environment, so it's not guaranteed the Arrow libs will be provided at runtime.

**Review by jlowe — CHANGES_REQUESTED** [review]
If `LICENSE-bundled` is being deleted then the corresponding `<resource>` node needs to be removed in java/pom.xml

File: java/README.md
Comment (codereport) [reply java/README.md]: Done!

File: java/README.md
Comment (jlowe) [inline java/README.md:55]:
```suggestion
Pass in the cmake option `-DCUDF_USE_ARROW_STATIC=ON` so that Apache Arrow is
```

File: java/ci/Dockerfile.centos7
Comment (jlowe) [inline java/ci/Dockerfile.centos7:34]: This entire RUN statement should be removed.  It's only for downloading and building boost from source (to get PIC object files even in a static lib).

**Review by jlowe — CHANGES_REQUESTED** [review]

**Review by ajschmidt8 — APPROVED** [review]

**Review by jlowe — APPROVED** [review]

**Comment by jrhemstad** [comment]
rerun tests

**Comment by harrism** [comment]
```
07:16:12 FAILED: CMakeFiles/cudf.dir/src/jit/cache.cpp.o 
07:16:13 ccache /usr/bin/g++ -DCUDF_VERSION=0.20.0 -DCUFILE_FOUND -DJITIFY_PRINT_LOG=0 -DJITIFY_USE_CACHE -DSPDLOG_ACTIVE_LEVEL=SPDLOG_LEVEL_INFO -DTHRUST_DEVICE_SYSTEM=THRUST_DEVICE_SYSTEM_CUDA -DTHRUST_HOST_SYSTEM=THRUST_HOST_SYSTEM_CPP -Dcudf_EXPORTS -I_deps/jitify-src -I_deps/libcudacxx-src/include -I../include -Iinclude -I../src -I_deps/thrust-src -I_deps/thrust-src/dependencies/cub -isystem $PREFIX/include -isystem $BUILD_PREFIX/include -isystem /usr/local/cuda/include -isystem /usr/local/cuda/lib64 -O3 -DNDEBUG -fPIC -Wall -Werror -Wno-unknown-pragmas -Wno-error=deprecated-declarations -Wno-deprecated-declarations -pthread -std=gnu++1z -MD -MT CMakeFiles/cudf.dir/src/jit/cache.cpp.o -MF CMakeFiles/cudf.dir/src/jit/cache.cpp.o.d -o CMakeFiles/cudf.dir/src/jit/cache.cpp.o -c ../src/jit/cache.cpp
07:16:13 ../src/jit/cache.cpp:22:10: fatal error: filesystem: No such file or directory
07:16:13  #include <filesystem>
07:16:13           ^~~~~~~~~~~~
07:16:13 compilation terminated.
```

So it appears that gcc is not finding the `filesystem` header. What is weird to me is this: `-std=gnu++1z` instead of `-std=c++17`.

**Review by kkraus14 — APPROVED** [review]
cmake / conda lgtm

**Comment by codereport** [comment]
@gpucibot merge

**Pull request merged**
